rust/src/writer.rs

```diff
@@ -0,0 +1,146 @@
+//! The writer module contains the DeltaWriter and DeltaWriterError definitions and provides a
+//! higher level API for performing Delta transaction writes to a given Delta Table.
+//!
+//! Unlike the transaction API on DeltaTable, this higher level writer will also write out the
+//! parquet files
+
+use arrow::record_batch::RecordBatch;
+use crate::{DeltaTableError, DeltaTransactionError};
+use parquet::basic::Compression;
+use parquet::file::properties::WriterProperties;
+use parquet::file::writer::InMemoryWriteableCursor;
+use parquet::arrow::ArrowWriter;
+use serde_json::Value;
+use std::convert::TryFrom;
+use std::sync::Arc;
+
+struct ParquetBuffer {
+    writer: ArrowWriter<InMemoryWriteableCursor>,
+    cursor: InMemoryWriteableCursor,
+}
+
+impl ParquetBuffer {
+    fn try_new(schema: arrow::datatypes::SchemaRef) -> Result<Self, DeltaTableError> {
+        // Initialize writer properties for the underlying arrow writer
+        let writer_properties = WriterProperties::builder()
+            // NOTE: Consider extracting config for writer properties and setting more than just compression
+            .set_compression(Compression::SNAPPY)
+            .build();
+
+        let cursor = InMemoryWriteableCursor::default();
+        let writer = ArrowWriter::try_new(
+            cursor.clone(),
+            schema,
+            Some(writer_properties),
+        )?;
+
+        Ok(Self { writer, cursor })
+    }
+
+    fn write_batch(&mut self, batch: &RecordBatch) -> Result<(), DeltaTableError> {
+        self.writer.write(batch).map_err(|source| DeltaTableError::ParquetError { source })
+    }
+
+    fn data(&self) -> Vec<u8> {
+        self.cursor.data()
+    }
+
+    fn close(&mut self) -> Result<parquet_format::FileMetaData, DeltaTableError> {
+        self.writer.close().map_err(|source| DeltaTableError::ParquetError { source })
+    }
+}
+
+struct BufferedJSONWriter {
```

Comment:
Nice! I like this as a more conventional API to write to for clients.

```suggestion
pub struct BufferedJSONWriter {
```